lib/open_api_spex/cast_parameters.ex

```diff
@@ -7,39 +7,94 @@ defmodule OpenApiSpex.CastParameters do
   @spec cast(Plug.Conn.t(), Operation.t(), Components.t()) ::
           {:error, [Error.t()]} | {:ok, Conn.t()}
   def cast(conn, operation, components) do
-    # Taken together as a set, operation parameters are similar to an object schema type.
-    # Convert parameters to an object schema, then delegate to `Cast.Object.cast/1`
-
-    # Operation's parameters list may include references - resolving here
-    resolved_parameters =
-      Enum.map(operation.parameters, fn
-        ref = %Reference{} -> Reference.resolve_parameter(ref, components.parameters)
-        param = %Parameter{} -> param
-      end)
+    full_cast_result =
+      schemas_by_location(operation, components)
+      |> Enum.map(fn {location, schema} -> cast_location(location, schema, components, conn) end)
+      |> reduce_cast_results()
+
+    case full_cast_result do
+      {:ok, params} -> {:ok, %{conn | params: params}}
+      err -> err
+    end
+  end
+
+  defp get_params_by_location(conn, :query, _) do
+    Plug.Conn.fetch_query_params(conn).query_params
+  end
 
-    properties =
-      resolved_parameters
-      |> Enum.map(fn parameter -> {parameter.name, Parameter.schema(parameter)} end)
-      |> Map.new()
+  defp get_params_by_location(conn, :path, _) do
+    conn.path_params
+  end
+
+  defp get_params_by_location(conn, :cookie, _) do
+    Plug.Conn.fetch_cookies(conn).req_cookies
+  end
+
+  defp get_params_by_location(conn, :header, expected_names) do
+    conn.req_headers
+    |> Enum.filter(fn {key, _value} ->
+      Enum.member?(expected_names, String.downcase(key))
+    end)
+    |> Map.new()
+  end
 
-    required =
-      resolved_parameters
-      |> Enum.filter(& &1.required)
-      |> Enum.map(& &1.name)
+  defp create_location_schema(parameters) do
+    properties = Map.new(parameters, fn p -> {p.name, Parameter.schema(p)} end)
 
-    object_schema = %Schema{
+    %Schema{
       type: :object,
+      additionalProperties: false,
       properties: properties,
-      required: required,
-      additionalProperties: false
+      required: parameters |> Enum.filter(& &1.required) |> Enum.map(& &1.name)
     }
+    |> maybe_add_additional_properties()
+  end
+
+  defp schemas_by_location(operation, components) do
+    param_specs_by_location =
+      operation.parameters
+      |> Enum.map(fn
+        %Reference{} = ref -> Reference.resolve_parameter(ref, components.parameters)
+        %Parameter{} = param -> param
+      end)
+      |> Enum.group_by(& &1.in)
+
+    Map.new(param_specs_by_location, fn {location, parameters} ->
+      {location, create_location_schema(parameters)}
+    end)
+  end
+
+  defp cast_location(location, schema, components, conn) do
+    params = get_params_by_location(conn, location, Map.keys(schema.properties))
```

Comment:
Does `Map.keys(schema.properties)` returns a list of atoms?

When it's used in `defp get_params_by_location(conn, :header, expected_names)`, shouldn't it be a list of strings?

If so, shouldn't the tests catch this?

Reply:
You're absolutely right, the main reason why the test didn't go onto this is that the parameter "x-user-id" is in fact defined as string 

see `test/support/pet_controller.ex:100` (https://github.com/open-api-spex/open_api_spex/pull/185/files#diff-6dd9f32e53930f9134af2ca31da212f6R100)

Reply:
I've just changed it to `schema.properties |> Map.keys() |> Enum.map(&Atom.to_string/1)`. I hope this is the propper way, since this would provoke a `ArgumentError` whenever a parameter key is defined as a `string`

Would it be better to add a more "tolerant` way?

Reply:
I think that's fine. If the property key is not an atom, that's a programmer error, and a raised error is appropriate.